### PR TITLE
[Gekidou] [MM-39718] Add Create DM screen

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Add Create DM screen

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39718

#### Screenshots
Screenshot from 2022-01-21 16-59-12
Screenshot from 2022-01-21 16-59-24
Screenshot from 2022-01-21 16-59-52
Screenshot from 2022-01-21 17-00-44
Screenshot from 2022-01-21 17-00-53
Screenshot from 2022-01-21 17-01-51

#### Release Note
```release-note
None
```

### CODE REVIEW

**Comment by larkox** [comment]
New UI screenshots:
Screenshot from 2022-01-21 17-50-10
Screenshot from 2022-01-21 17-50-19

**Comment by enahum** [comment]
The UI looks weird in terms of the navigation bar, it looks like is taking more vertical space than it should

**Comment by larkox** [comment]
@enahum may have to do with my emulator. All the screens have that amount of "air" on the navigation bar. What can be noticed in these screenshots, is that there is a color mismatch between the "system area" and the "close button area". Not sure exactly where that should be solved.

**Comment by shazm** [comment]
It looks pretty good to me - but I think search does not like the auto created users in the test env:

https://user-images.githubusercontent.com/456511/152921425-6bc3817e-4792-4d9b-9006-5639855d725e.mov

**Comment by larkox** [comment]
@shazm Seems to be working on my local. My only guess is that there is something weird happening with the search on the server? (like elastic search not being up to date, or anything like that)

**Comment by shazm** [comment]
> @shazm Seems to be working on my local. My only guess is that there is something weird happening with the search on the server? (like elastic search not being up to date, or anything like that)

Oo good point - I'm testing against a local server instance, will try with `test.mattermost`.

**Comment by Rina-dsg** [comment]
1. Chips should be vertically centered.
Screenshot 2022-02-23 at 15 46 51

2. These 2 distances should be the same (12 pt)
Screenshot 2022-02-23 at 15 48 44

**Comment by larkox** [comment]
@shazm @enahum This is ready to review.

File: app/actions/remote/channel.ts [hunk 1]
```diff
@@ -547,6 +548,57 @@ export const switchToChannelByName = async (serverUrl: string, channelName: stri
...
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+
+    try {
+        const currentUserId = await queryCurrentUserId(operator.database);
+        const created = await client.createDirectChannel([userId, currentUserId]);
```
Comment (enahum) [inline app/actions/remote/channel.ts:566]: you need to set the display name to the DM channel

File: app/actions/remote/channel.ts [hunk 2]
```diff
@@ -583,6 +662,58 @@ export const fetchArchivedChannels = async (serverUrl: string, teamId: string, p
...
+    }
+
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+    try {
+        const created = await client.createGroupChannel(userIds);
```
Comment (enahum) [inline app/actions/remote/channel.ts:678]: you need to set the display name to the GM channel

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/components/profile_picture/status.tsx [hunk 3]
```diff
@@ -39,8 +39,8 @@ const Status = ({author, statusSize, statusStyle, theme}: Props) => {
         statusStyle,
         {borderRadius: statusSize / 2},
     ]), [statusStyle]);
-
-    if (author?.status && !author.isBot) {
+    const isBot = author && (('isBot' in author) ? author.isBot : author.is_bot);
```
Comment (enahum) [inline app/components/profile_picture/status.tsx:42]: not really part of the PR, but in app/components/profile_picture/index.tsx we have a call to get the user status, I wonder if we should skip it if the user is a bot.

File: app/screens/more_dms/more_dms.tsx [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';
+import {useIntl} from 'react-intl';
+import {Keyboard, View} from 'react-native';
+import {Navigation} from 'react-native-navigation';
+import {SafeAreaView} from 'react-native-safe-area-context';
+
+import {makeDirectChannel, makeGroupChannel} from '@actions/remote/channel';
+import {fetchProfiles, fetchProfilesInTeam, searchProfiles} from '@actions/remote/user';
+import CompassIcon from '@app/components/compass_icon';
+import {useServerUrl} from '@app/context/server';
```
Comment (enahum) [inline app/screens/more_dms/more_dms.tsx:13]:
```suggestion
import CompassIcon from '@components/compass_icon';
import {useServerUrl} from '@context/server';
```

and sort

File: app/screens/more_dms/more_dms.tsx [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+        Reflect.deleteProperty(newSelectedIds, id);
+
+        setSelectedIds(newSelectedIds);
+    }, [selectedIds]);
+
+    const createDirectChannel = useCallback(async (id: string): Promise<boolean> => {
+        const user = profiles.find((u) => u.id === id);
+
+        const displayName = displayUsername(user, intl.locale, teammateNameDisplay);
```
Comment (enahum) [inline app/screens/more_dms/more_dms.tsx:159]: interesting.. if you have the displayName here I wonder if it can be passed as an optional argument to `makeDirectChannel` and then avoid having to do the `fetchMissingSidebarInfo` call...

thoughts ?

File: app/screens/more_dms/more_dms.tsx [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+    const createGroupChannel = useCallback(async (ids: string[]): Promise<boolean> => {
+        const result = await makeGroupChannel(serverUrl, ids);
+
+        if (result.error) {
+            alertErrorWithFallback(
+                intl,
+                result.error,
+                {
+                    id: t('mobile.open_gm.error'),
+                    defaultMessage: "We couldn't open a group message with those users. Please check your connection and try again.",
```
Comment (enahum) [inline app/screens/more_dms/more_dms.tsx:189]: same with connection here. cc: @cwarnermm

File: app/screens/more_dms/more_dms.tsx [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+    }, [searchUsers, clearSearch]);
+
+    const updateNavigationButtons = useCallback(async (startEnabled: boolean) => {
+        const closeIcon = await CompassIcon.getImageSource('close', 24, theme.sidebarHeaderTextColor);
+        setButtons(componentId, {
+            leftButtons: [{
+                id: CLOSE_BUTTON,
+                icon: closeIcon,
+                testID: 'close.more_direct_messages.button',
+            }],
```
Comment (enahum) [inline app/screens/more_dms/more_dms.tsx:295]: the left button does not seem to change at any moment, perhaps best to open the modal and setting the left button there and only update the right button here? 0/5

File: app/screens/more_dms/selected_users/index.tsx [hunk 8]
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useMemo} from 'react';
+import {View} from 'react-native';
+
+import FormattedText from '@components/formatted_text';
+import {useTheme} from '@context/theme';
+import SelectedUser from '@screens/more_dms/selected_users/selected_user';
```
Comment (enahum) [inline app/screens/more_dms/selected_users/index.tsx:9]:
```suggestion
import SelectedUser from './selected_user';
```

**Review by enahum — CHANGES_REQUESTED** [review]
No need to change the strings in non `en.json` localization files.

File: app/actions/remote/user.ts [hunk 9]
```diff
@@ -257,9 +258,15 @@ export const fetchUsersByIds = async (serverUrl: string, userIds: string[], fetc
     }
 
     try {
-        const currentUserId = await queryCurrentUserId(operator.database);
-        const exisingUsers = await queryUsersById(operator.database, userIds);
-        const usersToLoad = userIds.filter((id) => (id !== currentUserId && !exisingUsers.find((u) => u.id === id)));
+        const currentUser = await queryCurrentUser(operator.database);
```
[the remark this answers is not in the document]
Comment (larkox) [reply app/actions/remote/user.ts:261]: We want to pass the whole current user in the exiting users result.

File: app/components/channel_list/header/header.tsx [hunk 10]
```diff
@@ -101,7 +101,7 @@ const ChannelListHeader = ({canCreateChannels, canJoinChannels, displayName, ico
         bottomSheet({
             closeButtonId,
             renderContent,
-            snapPoints: [(items * ITEM_HEIGHT) + (insets.bottom * 2), 10],
+            snapPoints: [((items + 1) * ITEM_HEIGHT) + (insets.bottom * 2), 10],
```
[the remark this answers is not in the document]
Comment (larkox) [reply app/components/channel_list/header/header.tsx:104]: Not sure, but I will take your word for it.

File: app/screens/more_dms/more_dms.tsx [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+        if (idsToUse.length === 0) {
+            success = false;
+        } else if (idsToUse.length > 1) {
+            success = await createGroupChannel(idsToUse);
+        } else {
+            success = await createDirectChannel(idsToUse[0]);
+        }
+
+        if (success) {
+            requestAnimationFrame(() => {
```
[the remark this answers is not in the document]
Comment (larkox) [reply app/screens/more_dms/more_dms.tsx:215]: I guess not.

**Review by enahum — APPROVED** [review]
